docs/developer_guidelines.rst

```diff
@@ -41,9 +40,92 @@ Blocks:
+each argument given, because
+
+    When I see a bird that walks like a duck and swims like a duck and quacks
+    like a duck, I call that bird a duck.
+
+    -- James Whitcomb Riley
+
+If you need to perform some sort of input validation, don't use ``assert``
+statements. Raise a ``ValueError`` instead. ``assert`` statements should
+only be used for sanity tests i.e. they *should* never be triggered, unless
+there is a bug in the code.
+
+.. _duck typing: https://en.wikipedia.org/wiki/Duck_typing
+
+Abstract classes
+~~~~~~~~~~~~~~~~
+If a class is an `abstract base class`_, use Python's |abc|_ to mark it as such.
+
+.. code-block:: python
+
+   from abc import ABCMeta
+   from six import add_metaclass
+   @add_metaclass(ABCMeta)
+   class Abstract(object):
+       pass
+
+Our documentation generator (Sphinx_ with the autodoc_ extension, running on
+`Read the Docs`_) doesn't recognize classes which inherit the ``ABCMeta``
+metaclass as abstract and will try to instantiate them, causing errors when
+building documentation. To prevent this, make sure to always use the
+``add_metaclass`` decorator, regardless of the parent.
+
+.. _abstract base class: https://en.wikipedia.org/wiki/Class_%28computer_programming%29#Abstract_and_concrete
+.. |abc| replace:: ``abc``
+.. _abc: https://docs.python.org/3/library/abc.html
+.. _Sphinx: http://sphinx-doc.org/
+.. _autodoc: http://sphinx-doc.org/ext/autodoc.html
+.. _Read the Docs: https://readthedocs.org/
+
+Python 2 and 3
+~~~~~~~~~~~~~~
+Blocks aims to be both Python 2 and Python 3 compliant using a single code-base,
+without using 2to3_. There are many online resources which discuss the writing
+compatible code. For a quick overview see `the cheatsheet from Python
+Charmers`_. For non-trivial cases, we use the six_ compatibility library.
+
+Documentation should be written to be Python 3 compliant.
+
+.. _2to3: https://docs.python.org/2/library/2to3.html
+.. _the cheatsheet from Python Charmers: http://python-future.org/compatible_idioms.html
+.. _six: https://pythonhosted.org/six/
+
+Reraising exceptions
+~~~~~~~~~~~~~~~~~~~~
+When catching exceptions, use the :func:`~blocks.utils.reraise_as` function to
+reraise the exception (optionally with a new message or as a different type).
+Not doing so `clobbers the original traceback`_, making it impossible to use
+``pdb`` to debug the problems.
+
+.. _clobbers the original traceback: http://www.ianbicking.org/blog/2007/09/re-raising-exceptions.html
+
+Serialization
+~~~~~~~~~~~~~
+To ensure the reproducibility of scientific experiments Blocks tries to make
+sure that stopping and resuming training doesn't affect the final results. In
+order to do so it takes a radical approach, serializing the entire training
+state using dill_ (an extension of Python's native pickle_). Some things cannot
+be pickled, so their use should be avoided:
+
+* Generators
+* Dynamically generated classes (*possible but complicated*)
```

Comment:
Unfortunately we have to add list iterators to this list of shame. And I just guess all other standard container iterators.